### PR TITLE
signature minter

### PR DESCRIPTION
**Pull request by oveddan**
code for a signature minter.

We use a unique hash of the minting parameters, and an auth registry for allowed signers, for security purposes.

Design influenced by [ThirdWeb's SigntureMinter](https://github.com/thirdweb-dev/contracts/blob/cfe8050c06082d99fb89dee8003526124303edb0/contracts/extension/SignatureMintERC721Upgradeable.sol#L11)
Signer must keep track of a random bytes32 that must be unique for a signture.

### CODE REVIEW

File: src/minters/signature/ZoraSignatureMinterStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+/// @title ZoraSignatureMinterStrategy
+/// @notice Mints tokens based on signature created by an authorized signer
+/// @author @oveddan
+contract ZoraSignatureMinterStrategy is Enjoy, SaleStrategy, LimitedMintPerAddress, EIP712 {
+    using SaleCommandHelper for ICreatorCommands.CommandSet;
+
+    /// @notice General signatue sale settings
+    struct SalesConfig {
+        IAuthRegistry authorizedSignatureCreators;
```
[the remark this answers is not in the document]
Comment (iainnash) [reply src/minters/signature/ZoraSignatureMinterStrategy.sol:52]: thoughts on making this optional and then have a fallback for when we don't need to set both of these

[the remark this answers is not in the document]
Comment (oveddan) [reply src/minters/signature/ZoraSignatureMinterStrategy.sol:52]: by hardcode it you mean set it once in the constructor, and have it be global among all tokens and contracts?

File: src/minters/signature/ZoraSignatureMinterStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    function requestMint(
+        address,
+        uint256 tokenId,
+        uint256 quantity,
+        uint256 ethValueSent,
+        bytes calldata minterArguments
+    ) external returns (ICreatorCommands.CommandSet memory) {
+        address target = msg.sender;
+        // these arguments are what don't fit into the standard requestMint Args
+        (bytes32 uid, uint256 pricePerToken, uint256 expiration, address mintTo, bytes memory signature) = abi.decode(
```
[the remark this answers is not in the document]
Comment (oveddan) [reply src/minters/signature/ZoraSignatureMinterStrategy.sol:109]: yeah i supposed nonce can work - but nonce implies autoincrementing correct? i guess we can auto-increment it off-chain.

File: src/minters/signature/ZoraSignatureMinterStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+        address signer = _recover(target, tokenId, uid, quantity, pricePerToken, expiration, mintTo, signature);
+
+        // do we need this setting to be there for each token, or just be the same across the board?
+        if (signer == address(0) || !isAuthorizedToSign(signer, target, tokenId)) {
+            revert InvalidSignature();
+        }
+
+        // do we need this to be also unique per signer?
+        if (minted[target][uid]) {
```
[the remark this answers is not in the document]
Comment (oveddan) [reply src/minters/signature/ZoraSignatureMinterStrategy.sol:122]: regarding rotating the signer addresses, if we rotate out a signer, that would make previously created signatures by it invalid.  is that the desired behavior?

File: src/minters/signature/ZoraSignatureMinterStrategy.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (oveddan) [reply src/minters/signature/ZoraSignatureMinterStrategy.sol]: done - renamed from uid -> nonce

File: src/minters/signature/ZoraSignatureMinterStrategy.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (oveddan) [reply src/minters/signature/ZoraSignatureMinterStrategy.sol:52]: per discussion - there will be one config per contract. the fundsRecipient will be part of the signature

File: test/minters/signature/ZoraCreatorSignatureSaleStrategy.t.sol [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+pragma solidity 0.8.17;
+
+import "forge-std/Test.sol";
+import {ZoraCreator1155Impl} from "../../../src/nft/ZoraCreator1155Impl.sol";
+import {Zora1155} from "../../../src/proxies/Zora1155.sol";
+import {IZoraCreator1155} from "../../../src/interfaces/IZoraCreator1155.sol";
+import {IMinter1155} from "../../../src/interfaces/IMinter1155.sol";
+import {ICreatorRoyaltiesControl} from "../../../src/interfaces/ICreatorRoyaltiesControl.sol";
+import {IZoraCreator1155Factory} from "../../../src/interfaces/IZoraCreator1155Factory.sol";
+import {ILimitedMintPerAddress} from "../../../src/interfaces/ILimitedMintPerAddress.sol";
```
Comment (iainnash) [inline test/minters/signature/ZoraCreatorSignatureSaleStrategy.t.sol:11]: I don't think we need `LimitedMintPerAddress`.

File: src/minters/signature/ZoraSignatureMinterStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+    /// @notice General signatue sale settings
+    struct SalesConfig {
+        /// @notice Registry that decides which account is authorized to sign signatures
+        /// to mint for the sale
+        IReadableAuthRegistry authorizedSignatureCreators;
+    }
+
+    // target -> settings
+    mapping(address => SalesConfig) signatureSaleSettings;
```
Comment (iainnash) [inline src/minters/signature/ZoraSignatureMinterStrategy.sol:55]: What's the gas savings of hardcoding a default authorized registry when none is explicitly set?

Comment (iainnash) [reply src/minters/signature/ZoraSignatureMinterStrategy.sol:55]: Another approach that could be useful is if the address is not a smart contract just use that address as the signer address so individual wallets can use this config without deploying the registry.

File: src/minters/signature/ZoraSignatureMinterStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        // validate that proper value was sent
+        if (quantity * mintRequestCalldata.pricePerToken != ethValueSent) {
+            revert WrongValueSent(quantity * mintRequestCalldata.pricePerToken, ethValueSent);
+        }
+
+        return _executeMintAndTransferFunds(target, tokenId, quantity, mintRequestCalldata.mintTo, ethValueSent, mintRequestCalldata.fundsRecipient);
+    }
+
+    /// Helper utility to encode additional arguments needed to send to mint
+    function encodeMinterArgumets(MintRequestCallData calldata mintRequestCalldata) external pure returns (bytes memory) {
```
Comment (iainnash) [inline src/minters/signature/ZoraSignatureMinterStrategy.sol:159]: nit: `encodeMinterArgume_n_ts` missing a n

File: package.json [hunk 7]
```diff
@@ -13,10 +13,10 @@
     "prettier": "prettier --write 'src/**/*.sol' 'test/**/*.sol' 'package/**/*.ts' 'wagmi.config.ts'",
     "coverage": "forge coverage --report lcov",
     "write-gas-report": "forge test --gas-report > gasreport.ansi",
-    "prepack": "node script/copy-deployed-contracts.mjs && yarn run wagmi && yarn bundle-configs && yarn build-ts && yarn copy-addresses-and-configs",
+    "prepack": "node script/copy-deployed-contracts.mjs && pnpm wagmi && pnpm bundle-configs && pnpm build-ts && pnpm copy-addresses-and-configs",
```
Comment (iainnash) [inline package.json:16]: lgtm

File: src/authRegistry/AuthRegistry.sol [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+pragma solidity 0.8.17;
+
+import {IReadableAuthRegistry} from "../interfaces/IAuthRegistry.sol";
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+
+contract AuthRegistry is IReadableAuthRegistry, Ownable2Step {
+    mapping(address => bool) public override isAuthorized;
+
+    function setAuthorized(address account, bool authorized) external onlyOwner {
+        isAuthorized[account] = authorized;
```
[the remark this answers is not in the document]
Comment (jgeary) [reply src/authRegistry/AuthRegistry.sol:13]: done

File: src/minters/signature/ZoraSignatureMinterStrategy.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgeary) [reply src/minters/signature/ZoraSignatureMinterStrategy.sol]: hey @iainnash @oveddan im struggling to view the whole outdated diff here possible bc of my rebase, has this comment been addressed?

[the remark this answers is not in the document]
Comment (jgeary) [reply src/minters/signature/ZoraSignatureMinterStrategy.sol]: resolved over a call, keeping this how it is

File: package.json (hunk 7 above)
Comment (iainnash) [inline package.json:16]: was this merged back into another diff? if so would be good to rebase.

File: src/minters/signature/ZoraCreatorSignatureMinterStrategy.sol [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+    error Expired(uint256 expiration);
+
+    bytes32 constant REQUEST_MINT_TYPEHASH =
+        keccak256(
+            "requestMint(address target,uint256 tokenId,bytes32 nonce,uint256 quantity,uint256 pricePerToken,uint256 expiration,address mintTo,address fundRecipient)"
+        );
+
+    /// @notice ContractURI for contract information with the strategy
+    function contractURI() external pure override returns (string memory) {
+        return "https://github.com/ourzora/zora-1155-contracts/";
```
Comment (oveddan) [inline src/minters/signature/ZoraCreatorSignatureMinterStrategy.sol:77]: is this the correct url, since it's a private repo?

Comment (jgeary) [reply src/minters/signature/ZoraCreatorSignatureMinterStrategy.sol:77]: yes

**Comment by iainnash** [comment]
the 1155 repo is public

File: src/minters/signature/ZoraCreatorSignatureMinterStrategy.sol [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+    struct SalesConfig {
+        /// @notice Registry that decides which account is authorized to sign signatures
+        /// to mint for the sale
+        IReadableAuthRegistry authorizedSignatureCreators;
+    }
+
+    // target -> settings
+    mapping(address => SalesConfig) signatureSaleSettings;
+    // target contract -> unique nonce -> if has been minted already
+    mapping(address => mapping(bytes32 => bool)) private minted;
```
Comment (oveddan) [inline src/minters/signature/ZoraCreatorSignatureMinterStrategy.sol:57]: looking at [this tweet](https://twitter.com/danielvf/status/1675825631240519680), it's great we are **not** doing signature used by signature hash.

**Comment by iainnash** [comment]
Closing for now.

**Pull request closed without merging**